### ISSUE TITLE
Bug: Mocking varargs on 1.9

### ISSUE THREAD

**Issue by darrenhaken** [mockk/mockk#224]
I'm attempting to use a Mock against a Java SDK which has a vararg final argument.
However, I receive an error that no answer was found. I have seen the recently closed bug saying this bug was fixed in Mockk 1.8 but I'm using 1.9. 

#### Context
* MockK version: 1.9
* OS: Mac Mojava
* Kotlin version: 1.3.11
* JDK version: 1.8
* JUnit version: 5.3.2
* Type of test: unit test

#### Stack trace
```
io.mockk.MockKException: no answer found for: Table(#5).extract(AVRO, gcs://test_zone, [null])

	at io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:99)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:51)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:54)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:272)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:25)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:20)
	at com.google.cloud.bigquery.Table.extract(Table.java:456)
	at co.uk.autotrader.etl.bigquery.BigQueryStaging.exportStagingTableToGcs(BigQueryStaging.kt:27)
	at co.uk.autotrader.etl.bigquery.BigQueryStaging.prepareAvroFiles(BigQueryStaging.kt:11)
	at co.uk.autotrader.etl.bigquery.BigQueryStagingTest$SucceedToUnloadToGCS.write a partition of the source dataset into a BigQuery staging table(BigQueryStagingTest.kt:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:532)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:171)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:167)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:114)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:59)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$4(NodeTestTask.java:108)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:98)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:74)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$4(NodeTestTask.java:112)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:98)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:74)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$4(NodeTestTask.java:112)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:98)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:74)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$4(NodeTestTask.java:112)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:98)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:74)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

Jan 11, 2019 6:24:04 PM org.junit.jupiter.engine.descriptor.TestInstanceLifecycleUtils getDefaultTestInstanceLifecycle
```

#### Minimal reproducible code (the gist of this issue)

```
testImplementation "io.mockk:mockk:1.9"
    testCompile (
        "org.junit.jupiter:junit-jupiter-api:5.3.2",
        "org.assertj:assertj-core:3.11.1",
    )

    testRuntime "org.junit.jupiter:junit-jupiter-engine:5.3.2"

```

```kotlin
import com.google.cloud.RetryOption
import com.google.cloud.bigquery.*
import io.mockk.*
import io.mockk.junit5.MockKExtension
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.*
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class BigQueryStagingTest {

   @Nested
    inner class SucceedToUnloadToGCS {
        @Test
        fun `write a partition of the source dataset into a BigQuery staging table`() {
            BigQueryStaging.prepareAvroFiles(stagingJobConfig, bigQuery, stagingJobInfo)
            verify { bigQuery.create(stagingJobInfo) }
            confirmVerified(bigQuery)
        }
    }

private fun mockBigQueryJob() {
        every { bigQuery.create(stagingJobInfo) } returns stagingJob
        every { stagingJob.waitFor(ofType(RetryOption::class), ofType(RetryOption::class)) } returns stagingJob
        every { stagingJob.status } returns stagingJobStatus

        every { stagingJobConfig.stagingTableId } returns stagingTableId
        every { bigQuery.getTable(stagingTableId) } returns stagingTable
        every { stagingJobConfig.bigQueryUnloadLocation } returns "gcs://test_zone"

// *********************** This line is the one that I cannot get working which expects varargs****        
every { stagingTable.extract("AVRO", stagingJobConfig.bigQueryUnloadLocation, *arrayOf(any())) } returns extractJob
        every { extractJob.waitFor(ofType(RetryOption::class), ofType(RetryOption::class)) } returns extractJob
        every { extractJob.status } returns extractJobStatus
    }
}
```

```kotlin
// This is the production code line
val job = table.extract(FormatOptions.avro().type, jobConfig.bigQueryUnloadLocation)!!

```

**Comment by oshamahue** [comment]
I have done some tests and found out you cannot have a matcher that would match any number of args, so you should have something like this to make it
`every { stagingTable.extract("AVRO", stagingJobConfig.bigQueryUnloadLocation, null} `

**Comment by streetsofboston** [comment]
Mocking functions that have a variable-arguments parameter,`vararg` (Kotlin) or `...`(Java), is easy to do on Mockito but harder on MockK. 

Using MockK, the workaround is to mock out the function with as many parameters is necessary for your unit-test to run correctly. E.g.

```
    private val mockContext: Context = mockk(relaxed = true) {
        every { getString(any()) } answers {
            it.invocation.args.asString()
        }

        every { getString(any(), any()) } answers {
            it.invocation.args.asString()
        }

        every { getString(any(), any(), any()) } answers {
            it.invocation.args.asString()
        }

        every { getString(any(), any(), any(), any()) } answers {
            it.invocation.args.asString()
        }

        every { getString(any(), any(), any(), any(), any()) } answers {
            it.invocation.args.asString()
        }
    }
    ...
    private fun List<Any?>.asString(): String ...
    ...
```
where `getString` has two overloaded definitions in the Java `Context` class:
`public final String getString(int resId)` and `public final String getString(int resId, Object... formatArgs)`. The issue is about the latter one.

This workaround can get cumbersome and it would be great if this could be fixed.

Also, but this is not a real issue, the arguments-list (`it.invocation.args`) on MockK represents the arguments of the above `getString` example always as a list of *two* items (one `Int` and one `Array<Any?>`), while on Mockito this list has the size of the actual number of arguments provided during the call.

**Comment by oleksiyp** [comment]
Will check. It is great you report problems like that as this allow to focus on real problems

**Comment by oleksiyp** [comment]
A bunch of new matchers introduced to deal with `vararg`. Hopefully, this will enough to cover needs. Now it is a challenge to write good docs

**Comment by oleksiyp** [comment]
Released [1.9.1](https://github.com/mockk/mockk/releases/tag/1.9.1). Please try it out and close the ticket if everything is fine. Official docs will be updated a little bit later
